include/event_manager/event_manager_priv.h

```diff
@@ -0,0 +1,304 @@
+
+
+#define _EVENT_TYPE_DECLARE_COMMON(ename)				\
+	extern Z_DECL_ALIGN(struct event_type) _CONCAT(__event_type_, ename);		\
+	_EVENT_CASTER_FN(ename);					\
+	_EVENT_TYPECHECK_FN(ename)
+
+
+#define _EVENT_TYPE_DECLARE(ename)					\
+	_EVENT_TYPE_DECLARE_COMMON(ename);				\
+	_EVENT_ALLOCATOR_FN(ename)
+
+
+#define _EVENT_TYPE_DYNDATA_DECLARE(ename)				\
+	_EVENT_TYPE_DECLARE_COMMON(ename);				\
+	_EVENT_ALLOCATOR_DYNDATA_FN(ename)
+
+
+#define _EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, trace_data_pointer)		\
+	_EVENT_SUBSCRIBERS_ARRAY_TAGS(ename);						\
+	STRUCT_SECTION_ITERABLE(event_type, _CONCAT(__event_type_, ename)) = {		\
+		.name            = STRINGIFY(ename),					\
+		.subs_start      = _EVENT_SUBSCRIBERS_START_TAG(ename),			\
+		.subs_stop       = _EVENT_SUBSCRIBERS_END_TAG(ename),			\
+		.init_log_enable = init_log_en,						\
+		.log_event       = (IS_ENABLED(CONFIG_LOG) ? (log_fn) : (NULL)),	\
+		.trace_data      = (IS_ENABLED(CONFIG_EVENT_MANAGER_TRACE_EVENT_DATA) ?	\
+					(trace_data_pointer) : (NULL)),			\
+	}
+
+
+/**
+ * @brief Bitmask indicating event is displayed.
+ */
+struct event_manager_event_display_bm {
+	ATOMIC_DEFINE(flags, CONFIG_EVENT_MANAGER_MAX_EVENT_CNT);
+};
+
+extern struct event_manager_event_display_bm _event_manager_event_display_bm;
+
+
+/** @brief Event header.
+ *
+ * When defining an event structure, the event header
+ * must be placed as the first field.
+ */
+struct event_header {
+	/** Linked list node used to chain events. */
+	sys_snode_t node;
+
+	/** Pointer to the event type object. */
+	const struct event_type *type_id;
+};
+
+
+/** @brief Dynamic event data.
+ *
+ * When defining an event structure, the dynamic event data
+ * must be placed as the last field.
+ */
+struct event_dyndata {
+	/** Size of the dynamic data. */
+	size_t size;
+
+	/** Dynamic data. */
+	uint8_t data[0];
+};
+
+
+/** @brief Event listener.
+ *
+ * All event listeners must be defined using @ref EVENT_LISTENER.
+ */
+struct event_listener {
+	/** Name of this listener. */
+	const char *name;
+
+	/** Pointer to the function that is called when an event is handled.
+	 * The function should return true to consume the event, which means that the event is
+	 * not propagated to further listeners, or false, otherwise.
+	 */
+	bool (*notification)(const struct event_header *eh);
+};
+
+
+/** @brief Event subscriber.
+ */
+struct event_subscriber {
+	/** Pointer to the listener. */
+	const struct event_listener *listener;
+};
+
+
+/** @brief Event type.
+ */
+struct event_type {
+	/** Event name. */
+	const char			*name;
+
+	/** Pointer to the array of subscribers. */
+	const struct event_subscriber	*subs_start;
+
+	/** Pointer to the element directly after the array of subscribers. */
+	const struct event_subscriber	*subs_stop;
+
+	/** Bool indicating if the event is logged by default. */
+	bool init_log_enable;
+
+	/** Function to log data from this event. */
+	int (*log_event)(const struct event_header *eh, char *buf,
```

Comment:
Lets assume that we would like to use only new logging method using `EVENT_MAANGER_LOG` macro in zephyr. Legacy method is only for backward compatibility in NCS. I would propose following approach: 
- Deprecate legacy logging method (the one less efficient using snprintf)
- Create here two fields with two typedefs. New function will be `void (*log_event_func_t)(const struct event_header *eh)`
- Use `_Generic` keyword do detect type of function provided by the user in the macro and assign to one field and leave another as null
- in processing you pick method based on which pointer is not null.

If `_Generic` keyword is not supported by the compiler legacy method cannot be used and that should be fine because it will only be used in NCS and all cortex_m compilers that we are aware of support `_Generic`.

Detection of `_Generic` support: https://github.com/zephyrproject-rtos/zephyr/blob/731241f8d049924591cf4f46dda26ddd0782d8c7/include/sys/cbprintf.h#L28

Examples of `_Generic` use: https://github.com/zephyrproject-rtos/zephyr/blob/731241f8d049924591cf4f46dda26ddd0782d8c7/include/sys/cbprintf_internal.h#L91